gramps_webapi/api/resources/emit.py

```diff
@@ -0,0 +1,94 @@
+"""Gramps Json Encoder."""
+
+import inspect
+
+import gramps.gen.lib as lib
+from flask import Response
+from flask.json import JSONEncoder
+from gramps.gen.db import DbBookmarks
+
+PRIMARY_CLASS_MAP = {
+    "Person": lib.Person,
+    "Family": lib.Family,
+    "Event": lib.Event,
+    "Place": lib.Place,
+    "Source": lib.Source,
+    "Citation": lib.Citation,
+    "Repository": lib.Repository,
+    "Media": lib.Media,
+    "Note": lib.Note,
+    "Tag": lib.Tag,
+}
+
+
+class GrampsJSONEncoder(JSONEncoder):
+    """Customizes Gramps Web API output."""
+
+    def __init__(self):
+        """Initialize class."""
+        JSONEncoder.__init__(self)
+        self.sort_keys = True
+        self.ensure_ascii = False
+        self.strip_empty_keys = False
+        self.filter_only_keys = []
+        self.filter_skip_keys = []
+        self.gramps_classes = [
+            getattr(lib, key) for key, value in inspect.getmembers(lib, inspect.isclass)
+        ]
+
+    def response(self, payload):
+        """Prepare response."""
+        return Response(
+            response=self.encode(payload),
+            status=200,
+            mimetype="application/json",
+        )
+
+    def api_filter(self, obj):
+        """Filter data for a Gramps object."""
+        data = {}
+        filter = False
+        try:
+            if isinstance(obj, PRIMARY_CLASS_MAP[self.gramps_class_name]):
+                filter = True
+        except:
+            pass
+        for key, value in obj.__dict__.items():
+            if filter:
+                if self.filter_only_keys != [] and key not in self.filter_only_keys:
+                    continue
+                if self.filter_skip_keys != [] and key in self.filter_skip_keys:
+                    continue
+            if key.startswith("_"):
+                key = key[2 + key.find("__") :]
+            if self.strip_empty_keys:
+                if isinstance(value, lib.GrampsType):
+                    data[key] = str(value)
+                elif isinstance(value, lib.StyledText):
+                    data[key] = str(value)
+                elif isinstance(value, lib.PlaceName):
+                    data[key] = {
+                        "date": value.date,
+                        "lang": value.lang,
+                        "value": value.value,
+                    }
+                else:
+                    if value is not None and value != [] and value != {}:
+                        data[key] = value
+            else:
+                data[key] = value
+        return data
```

Comment:
If the value itself is a Gramps object the comparison to an empty list or dict will fail and throw an exception as the later have no value. This is sort of an attempted optimization but without careful review of the Gramps schema so some cases may be missed. It could be:

            if self.strip_empty_keys:
                for gramps_class in self.gramps_classes:
                    if isinstance(value, gramps_class):
                        data[key] = value
                        break
		if key not in data:
                    if value is not None and value != [] and value != {}:
                        data[key] = value
            else:
               data[key] = value

Thoughts?

Reply:
I still don't understand why the Gramps class instances are converted to strings (`data[key] = str(value)`) but *only* when `strip_empty_keys` is true. Why is that?

What I would find logical would be simply
```python
if not self.strip_empty_keys or not is_null(value):
    data[key] = value
```

and then
```python
def is_null(object) -> bool:
    ...
```
should somehow handle how to determine whether Gramps objects are null

Reply:
Oh! Actually isn't this more proper?

if not self.strip_empty_keys or value:
    data[key] = value

Reply:
I think this won't work as e.g. `0` evaluates as `False`. So you will discard all keys where the value is 0.

Reply:
In addition, I wouldn't know what `bool(some_gramps_object)` even does...

Reply:
Oh that's right, I forgot 0.  Need to tack that last test on the end. 

As I currently understand it bool of an object returns True if it exists unless the class implements a __bool__() or __len()__ method that happens to return otherwise. bool of [] or {} or None always returns False. So that seems to be a reasonable test to me.

Reply:
Ah but False might be the value of the key. Crumb buns.

Reply:
So this should do, am I still overlooking a case...
`            if not self.strip_empty_keys or value or int(value) == 0 or isinstance(value, bool):
		data[key] = value
`

Reply:
But `int` will fail very often, e.g. on lists or dicts.

I think it's cleaner if we just check explicitly what we want to remove: iterables with length zero and `None`; that's it. Right?

So I would do:

```python
if not self.strip_empty_keys or not is_null(value):
    data[key] = value

def is_null(value):
    if value is None:
        return True
    try:
        return len(value) == 0
    except AttributeError:  # object has no length
        pass
    return False
```

Reply:
Yes you are right, done. I did not know you could nest functions, that is neat!  So much to learn...